### DISCUSSION TITLE
Klick'r 3.1.0-beta01

### DISCUSSION THREAD

**Discussion by Nain57** [Nain57/Smart-AutoClicker#546] [Announcements]
> [!IMPORTANT]
> **This is a beta build.**
> It might be unstable so please backup your scenarios before updating to this version. It will be only availble on direct download here, and not on PlayStore/FDroid.

## New Action: Notifications

You can now send notification to your device as an action in Smart scenarios.
* Print a predefined text, a display the value of one of your counters
* 3 predifined alterting profile, and you can customize them in the notification settings of you device.
* Notifications are limited by the system, we can't spam them at autoclicker speed. To ensure Klick'r is not muted, notifications are batched and updated every 2s

![Capture d'écran 2024-11-17 153950](https://github.com/user-attachments/assets/8cc4875f-1d56-43ac-81a0-7cd6c95bfec6)

## New Click Action feature: Click Offset

When clicking on the detected condition, you can now define an offset from the condition center, allowing to click relatively to the detected condition position.

![Capture d'écran 2024-11-02 143822](https://github.com/user-attachments/assets/3ae961eb-b7f2-469b-b72e-223b192abc84)

## New event setting: Keep detecting

Something that has never been clear for Klick'r users is why image events have priorities and why only one event at their list is detected. By default, Klick'r stops the event list processing as soon as one event is fullfiled, and restart the processing from the first event as soon as a new screen frame is available.
Now, this is still the case, but the user can enable/disable it via a setting in each event, allowing more flow control and maybe more understanding of the app by the users.

![Capture d'écran 2024-11-17 154055](https://github.com/user-attachments/assets/7934ee03-af7a-4ab6-8dcb-c0ac7a514033)

## Misc

* MediaProjection permission can now be requested again if lost while a scenario is loaded
* Confirmation is now asked when closing  whitout saving a configuration dialog with user changes
* Scenario are now collapsible in the Scenario list
* Add a button to copy a scenario in the scenario list
* Reuse the previous screen frame if a new one is not available yet after a processing iteration (this might have some light impact on your scenario timings)

<hr /><em>This discussion was created from the release <a href='https://github.com/Nain57/Smart-AutoClicker/releases/tag/3.1.0-beta01'>Klick'r 3.1.0-beta01</a>.</em>

**Comment by Supreme-Darkness** [comment]
I am not sure if the keep executing option is working. What I did was very simple. 

1. I created event_1 and asked the software to detect image_1 which is on screen_1. Once image_1 is detected, i asked it to click on a specific place and wait for 2 seconds. Then click on another specific place and wait for 2 seconds and so on.

2. Everytime it clicked as per instructions in (1.) image_2 will show on screen_2.

3.. Then, I created an event_2 to detect image_2 on screen_2,  and i asked it to click on a specific position on screen_2 when image_2 is detected.

4. Event_2 will do the desired click and screen_1 will show again.

5. So ideally flow should be like, 
Event_1 does its first click, waits 2 seconds,
Image_2 on screen_2 pops up while event_1 is waiting.
Event_2 will see image_2 on screen_2 and it will click.
This will show screen_1 again.
2 seconds wait will be over and the event_1 will continue it's execution from where it left (click 2). 
then event_2 will again do it's click.

6. But in reality event_2 never executes and event_1 keeps clicking. 

What I am saying is, this programme stops detection for other image events when one event is running it's instructions. 

And even if it keeps checking and detecting for other image events while one event is running, it surely doesn't execute any instructions from another event as long as event 1 is running it's own instructions..

**Reply by Nain57** [reply]
Hi and thank your for your feedback.

This use case is verified by this unit test: https://github.com/Nain57/Smart-AutoClicker/blob/master/core/smart/processing/src/test/java/com/buzbuz/smartautoclicker/core/processing/tests/processor/ProcessingTests.kt#L219, and i can't reproduce it on my devices.

This is most likely an error in your configuration, your second condition might not be detected at all. To ensure correctness, you can:
* Use the test option in the image condition screen to chek if your condition is correctly detected.
* Use the test option in the event dialog if you combine multiple conditions in a single event to check if all conditions are detected together.
* Use the notification action as a verification means, sending a different notification everytime event1 & event2 are fulfilled.

If you still think this is a bug please open a new issue and provide the minimal reproduction scenario with it: https://github.com/Nain57/Smart-AutoClicker/issues/new?assignees=Nain57&labels=bug&projects=&template=bug_report.yml

**Comment by robertzPL** [comment]
New list scenarios has a bug. If name is to long the name display not correctly

**Reply by Nain57** [reply]
Thank you for your feedback, I will take a look at that for beta02
